### ISSUE TITLE
[TCL] Improper handling of [] and {} brackets

### ISSUE THREAD

**Issue by Nindaleth** [sublimehq/Packages#698]
I have isolated a few (separately testable) pieces of code that currently break syntax highlighting the hardest for us - this is reproducable on 3126. If there are any other issues, I don't have any such easy culprit at hand.
Generally, I would say that Tcl has a special handling of [] and {} brackets, highlighting of which is done in a bit different way than what `tclsh` does.

```
#- the formatting string %02d should be hightlighted as part of the string, it's not a bareword
  set dirname "${v_seagull_cfg_root}/seagull-[format "%02d" [expr $ctrlport - $v_seagull_ctrl_port]]"
#- list is a keyword that should be colored as that, setting empty list like this doesn't color properly
  set a [list]
#- the commands should be colored... at the very least the commas in quotes should not be "non-string"
  set res "[join [lrange [split $res ","] 0 end-1] ","] ..."
#- the odd number of quotes causes stringifying of the rest
  regexp {instance="?([^" \t]+)"?} $counter matchedstring instance; # comment
#- everything is colored as a part of a string after this one - I mean everything as in "the remaining 4k lines of the file"
#  that all look like one giant string, cannot be CTRL+R'd, etc.
  set stepquote [regsub -all {"} $line {""} ]
  set this_is_still_seen_as_string true
```
This is how I see it in reality:

![st_3126_tcl_hl](https://cloud.githubusercontent.com/assets/6383928/20524274/01e8e118-b0ba-11e6-8010-b1547c6fb57a.png)

In summary, it seems that (aside of the #645 regarding \\ character) the most impacting is highlighting issues within the {}-enclosed strings.

**Comment by wbond** [comment]
So from your example code, am I correct in assuming that TCL evaluates `[]` expressions inside of strings? Is there an escape character that prevents `[` from being the beginning of an expression inside of a string?

**Comment by Nindaleth** [comment]
Yes, that is exactly right, see this example done directly in tclsh:

```
% puts "One plus one is [expr 1+[expr 3-2]]"
One plus one is 2
% puts "Here's how to escape it: \[expr 1+1\]"
Here's how to escape it: [expr 1+1]
```

The {} brackets, on the other hand, behave a bit differently:

```
% puts "One plus one is [expr {1+[expr 3-2]}]"
One plus one is 2
% puts {1+[expr 3-2]}
1+[expr 3-2]
```

**Comment by wbond** [comment]
These are all fixed by 5a6462ff7dd943278e2c05c19007e410ab52bcc1, which will be part of the next dev build

**Comment by Nindaleth** [comment]
Thanks a lot!
